### PR TITLE
jasmine tests and extended array buffers util code

### PR DESCRIPTION
**Pull request by iislucas**
This is intended to help provide a template for writing tests in Typescript with Jasmine. 
- added `hexStringToArrayBuffer`
- added typescript definition interface file for `ArrayBuffers`
- added third_party/DefinitelyTyped directory with README
- added typescript DefinitelyTyped jasmine interface files to third_party
- added jasmine tests for `ArrayBuffers` code. 
- updated Gruntfile to run new tests

TESTED: grunt test

### CODE REVIEW

File: spec/common/arraybuffers_spec.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    var a1 = new Uint8Array(b1);
+    var a2 = new Uint8Array(b2);
+    if(a1.byteLength !== a2.byteLength) return false;
+    for(var i:number = 0; i < a1.byteLength; ++i) {
+      if(a1[i] !== a2[i]) return false;
+    }
+    return true;
+  }
+
+  describe("ArraBuffers <-> Hex Strings", function() {
```
Comment (trevj) [inline spec/common/arraybuffers_spec.ts]: typo: ArraBuffers -> ArrayBuffers

File: spec/common/arraybuffers_spec.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    it("arrayBufferEquality: emptyArray == emptyArray", function() {
```
Comment (trevj) [inline spec/common/arraybuffers_spec.ts:30]: Testing your tests, cool!

File: spec/common/arraybuffers_spec.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    it("Buffer -> Hex", function() {
+      expect(arrayBufferToHexString(emptyArray)).toEqual(emptyString);
+      expect(arrayBufferToHexString(array1)).toEqual(hexString1);
+      expect(arrayBufferToHexString(array2)).toEqual(hexString2);
+    });
+
+    it("Hex -> Buffer -> Hex = identity", function() {
+      expect(arrayBufferToHexString(hexStringToArrayBuffer(emptyString)))
+          .toEqual(emptyString);
+      expect(arrayBufferToHexString(hexStringToArrayBuffer(emptyString)))
```
Comment (trevj) [inline spec/common/arraybuffers_spec.ts:60]: Not sure this gives you anything? Well, OK, I see what you're doing. Cool.

File: spec/common/arraybuffers_spec.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+          .toEqual(hexString1);
+      expect(arrayBufferToHexString(hexStringToArrayBuffer(hexString2)))
+          .toEqual(hexString2);
+      expect(arrayBufferToHexString(hexStringToArrayBuffer(hexString1)))
+          .not.toEqual(hexString2);
+    });
+  });
+
+
+  describe("ArraBuffers <-> strings", function() {
```
Comment (trevj) [inline spec/common/arraybuffers_spec.ts]: same typo :-)

File: src/interfaces/arraybuffers.d.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+declare module ArrayBuffers {
```
Comment (trevj) [inline src/interfaces/arraybuffers.d.ts:1]: So, I'm curious what definding the interface here provides over just importing the implementation directly?

**Comment by trevj** [comment]
Nice! I like the arraybuffers tests. Thanks for adding those.

Regarding third_party, what's here is fine but I think perhaps we should investigate this:
https://www.npmjs.org/package/tsd

File: src/interfaces/arraybuffers.d.ts (hunk 5 above)
Comment (iislucas) [inline src/interfaces/arraybuffers.d.ts:1]: Useful for other files that assume arraybuffers are defined in the JS env, but don't get compiled in the same compilation command as arraybuffers themselves.  But mostly not used. (generated by `tsc -d`)

**Comment by iislucas** [comment]
+1 for using tsd. Added an issue for it (https://github.com/uProxy/socks-rtc/issues/60)

File: spec/common/arraybuffers_spec.ts (hunk 1 above)
Comment (iislucas) [inline spec/common/arraybuffers_spec.ts]: fixed

File: spec/common/arraybuffers_spec.ts (hunk 4 above)
Comment (iislucas) [inline spec/common/arraybuffers_spec.ts]: fixed.

File: spec/common/arraybuffers_spec.ts (hunk 3 above)
Comment (iislucas) [inline spec/common/arraybuffers_spec.ts:60]: Testing the things that should fail. :)

**Pull request merged**